Allow Per-department permissions

KludgeKML opened the pull request:
Adds the ability to mark a service as owned by one or more organisations, allowing users from those organisations that have been granted access to Local Links Manager in signon to view and edit links for those services. Other existing actions, and the ability to set owners, now only available with the new `GDS Editor` permission.

A menu option to Switch app has been added, consistent with Whitehall.

New tests are Request/System specs - we have not updated old Controller/Feature tests in this app, that has been captured as technical debt for updating later.

⚠️ **WARNING: merging PR without first updating existing GDS users of LLM with the new permission will result in them losing the ability to view local authorities and services. Update their permissions before merging!** ⚠️

- [x] Update permissions

https://trello.com/c/1eUXHHBz/22-restrict-access-to-a-specific-service, [Jira issue PNP-8560](https://gov-uk.atlassian.net/browse/PNP-8560)

## Screenshots

### View of the services page when logged in as a user (note main menu does not offer the broken links or council options)

![Screenshot 2024-08-14 at 11 24 26](https://github.com/user-attachments/assets/20e41b82-0c6e-437a-b192-8c85fac5736b)

### View of an owned service page when logged in as a user (note main menu does not offer the broken links or council options, update owners link is not available in sidebar)

![Screenshot 2024-08-14 at 11 24 34](https://github.com/user-attachments/assets/1692c82a-5ada-40d7-b03b-bd44a1082374)

### View of the edit link page for an owned service page when logged in as a user (note main menu does not offer the broken links or council options, delete option is not available)

![Screenshot 2024-08-14 at 11 24 43](https://github.com/user-attachments/assets/2cd65be1-dc32-42a2-a886-5c8326628057)

### View of an owned service page when logged in with GDS Editor permissions (update owners link is available in sidebar)

![Screenshot 2024-08-14 at 11 22 05](https://github.com/user-attachments/assets/3d402f83-6e9c-4bea-8b09-aafa0bf49240)

### New Update Owners form, available on when logged in with GDS Editor permissions 

![Screenshot 2024-08-14 at 11 22 19](https://github.com/user-attachments/assets/e43a36c2-6dac-490a-9fa4-a076ee72494a)



⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

KludgeKML committed:
Add ADR about department permissions

sairamya93 committed:
Add organisation_slugs to service record

KludgeKML committed:
Add documentation for new permission, feature

leenagupte commented on app/controllers/services_controller.rb:
```diff
@@ -18,6 +21,22 @@ def show
   end
 
+  def update_owner_form
+    raise GDS::SSO::PermissionDeniedError, "You do not have permission to view this page" unless gds_editor?
```
This is a new form? It and it's related bits (view, routes, etc) should really be split out into a separate commit.

KludgeKML replied:
I've updated the permissions to match the new pattern.

leenagupte reviewed (commented):
This is looking really good.

I especially like the use of system and request tests.

I've left a few inline comments, but there isn't anything major.

One thing I have noticed though, that Local links manager doesn't have a way to get users back to Sign-on to choose another app, e.g. a departmental publisher wanting to go into Whitehall.

See the "switch app" menu option in Whitehall:
![Screenshot 2024-08-13 at 14 13 54](https://github.com/user-attachments/assets/058bf59a-a050-4c0f-a514-f1176658d773)

Compared to Local links manager:
![Screenshot 2024-08-13 at 14 14 11](https://github.com/user-attachments/assets/cb497be1-03c8-45b7-8a01-f0ef8dca4ea5)

If we are opening this up to non-GDS people, we probably need to add something so that they don't get trapped.

KludgeKML commented on app/controllers/services_controller.rb:
```diff
@@ -3,10 +3,13 @@ class ServicesController < ApplicationController
-    @services = Service.enabled.order(broken_link_count: :desc)
-    raise "Missing Data" if @services.empty?
+    @services = services_for_user(current_user).enabled.order(broken_link_count: :desc)
+    # raise "Missing Data" if @services.empty?
```
No, it should be got rid of entirely (same with the test) - basically, it's perfectly legit for there to be no services returned to a given user now, and if the services all vanish entirely this shouldn't be the way we find out.

leenagupte replied:
👍  so it just needs to be deleted.

KludgeKML replied:
Done.

KludgeKML committed:
Update main menu

- Add Switch app option
- Add check so that without the GDS Editor permission,
  a user won't see the Broken Links or Councils links
- Add new System test to check this
- Update authentication support file to simplify it
- Include authentication by default in all test types
- Add infer test type to RSpec

Co-authored-by: Ramya Vidapanakal <ramya.vidapanakal@digital.cabinet-office.gov.uk>

KludgeKML committed:
Only show DELETE button to gds editors

- Add system test

KludgeKML committed:
Show owning department name in services list

Co-authored-by: Ramya Vidapanakal <ramya.vidapanakal@digital.cabinet-office.gov.uk>

KludgeKML committed:
Improve service page

- Add owners to info block
- Turn council link into plain text, since it
  is confusing having two action links in the
  table row anyway, and department users
  shouldn't be able to see them

Co-authored-by: Ramya Vidapanakal <ramya.vidapanakal@digital.cabinet-office.gov.uk>

KludgeKML committed:
Update existing tests

- Use login_as_gds_editor because all old tests
  just assume you have all powers.
- Add default organisation for factory user

Co-authored-by: Ramya Vidapanakal <ramya.vidapanakal@digital.cabinet-office.gov.uk>

KludgeKML commented on spec/support/authentication.rb:
```diff
@@ -1,4 +1,10 @@
+    allow_any_instance_of(ApplicationController).to receive(:current_user).and_return(user)
+  end
+
   def login_as(user)
```
Now removed, and this support file is simplified a bit.

KludgeKML commented on app/views/layouts/application.html.erb:
```diff
@@ -16,26 +16,29 @@
+    text: "Councils",
+    href: local_authorities_path(filter: %w[only_active]),
+    active: current_page?(local_authorities_path),
+  } if gds_editor?
```
Have done something similar.

KludgeKML commented on spec/support/authentication.rb:
```diff
@@ -11,8 +17,69 @@ def stub_user
+  end
+end
+
+RSpec.shared_examples "redirects non-GDS Editors to services page" do |path|
```
Shared examples now brought in the first time they're used.

KludgeKML commented on app/controllers/links_controller.rb:
```diff
@@ -1,6 +1,9 @@
 class LinksController < ApplicationController
   before_action :load_dependencies, only: %i[edit update destroy]
   before_action :set_back_url_before_post_request, only: %i[edit update destroy]
+  before_action :check_permissions, except: %i[edit update destroy]
```
I've made three more standard permission checking methods in the concern, so they're all handled as before actions now, and the names are consistent between controllers.

KludgeKML commented on app/controllers/links_controller.rb:
```diff
@@ -33,6 +36,8 @@ def update
   end
 
   def destroy
+    raise GDS::SSO::PermissionDeniedError, "You do not have permission to delete this link" unless gds_editor?
```
(See above)

KludgeKML commented on spec/requests/edit_link_page_spec.rb:
```diff
@@ -0,0 +1,100 @@
+    context "as a department user from the owning department" do
+      before { login_as_department_user(organisation_slug: owning_department) }
+
+      it "returns 403 Forbidden" do
```
Thanks! fixed.

KludgeKML commented on spec/requests/council_page_spec.rb:
```diff
@@ -0,0 +1,79 @@
+RSpec.describe "Council page" do
```
Added (in a different file, because it's on this controller but isn't a council page per se)

KludgeKML commented on app/presenters/links_table_presenter.rb:
```diff
@@ -15,7 +15,7 @@ def rows
         { text: "<span class=\"app-service-label\">#{title}</span>".html_safe },
         { text: "<span class=\"app-interaction-label\">#{link.interaction.label}</span>".html_safe },
-        { text: @view_context.link_to(link.local_authority.name, @view_context.local_authority_path(link.local_authority), class: "govuk-link") },
+        { text: link.local_authority.name },
```
Now removed in the same commit as the test

KludgeKML commented on spec/features/services/service_show_spec.rb:
```diff
@@ -43,10 +43,6 @@
   end
 
   describe "for each local authority" do
-    it "the Local Authority name is linked to the council specific page" do
```
Now test and code removed in same commit.

KludgeKML commented on spec/controllers/services_controller_spec.rb:
```diff
@@ -1,11 +1,11 @@
+    #     login_as_stub_user
+    #     expect { get :index }.to raise_error "Missing Data"
+    #   end
+    # end
```
Now removed.

leenagupte reviewed (commented):
Thanks for updating how permissions are checked so that they follow a consistent pattern 🎉 

I've resolved all the comments from the previous review. Thanks for fixing.🥇 

The only thing that's still a bit of a brain twister is the logic for permissions in the controllers. I've tried to think of ways to simplify the logic in inline comments, but it's tricky. I think it might be easier to follow if permissions are specifically applied to controller actions. i.e. actions a, b, and c require "gds_editor", d, e and f require "gds_editor" or "organisation", and if the action doesn't require any then, don't mention it. It might make the code a bit more verbose, but there wouldn't be any need to figure out what needs what.

I've also added a few other minor comments inline.

KludgeKML committed:
Use permissions in links controller

- Auto-redirect to services page if not a GDS Editor,
  or forbid if it's not a navigable page.
- Edit link can be accessed by an owner.
- Add request tests to enforce these checks.
- Add shared examples to make tests simpler.

Co-authored-by: Ramya Vidapanakal <ramya.vidapanakal@digital.cabinet-office.gov.uk>

KludgeKML committed:
Enforce permissions on councils pages

- Only GDS Editors can see local authority list, or
  use local authority endpoints.
- Add request tests

Co-authored-by: Ramya Vidapanakal <ramya.vidapanakal@digital.cabinet-office.gov.uk>

KludgeKML committed:
Enforce permissions on services page

- GDS Editors can see list of all services, department
  users can only see their own services.
- All actions on services can only be performed by
  GDS Editors or users from owning departments.
- It's now legit for the services page not to have any
  services appearing because the current user might
  not belong to a department that owns any yet, and
  for GDS Editors a catastrophic loss of data shouldn't
  be a thing that a user has to detect.
- Make the upload services CSV work like the same
  one in the services controller (redirect to the service
  on a good import, redirect to the form on a bad one)

Co-authored-by: Ramya Vidapanakal <ramya.vidapanakal@digital.cabinet-office.gov.uk>2

KludgeKML committed:
Add update_owners form

- Only GDS Editors can see this for the moment

Co-authored-by: Ramya Vidapanakal <ramya.vidapanakal@digital.cabinet-office.gov.uk>

KludgeKML commented on spec/requests/services_page_spec.rb:
```diff
@@ -13,4 +13,60 @@
+      it "updates the organisation slugs" do
+        patch "/services/aardvark-wardens/update-owner", params: { service: { organisation_slugs: "new-owner" } }
+
+        expect(service.reload.organisation_slugs).to eq(%w[new-owner])
```
Added

KludgeKML commented on spec/requests/services_page_spec.rb:
```diff
@@ -0,0 +1,16 @@
+RSpec.describe "Services page" do
```
Added

KludgeKML commented on spec/requests/bad_homepage_csv_spec.rb:
```diff
@@ -0,0 +1,3 @@
+RSpec.describe "Bad homepage CSV" do
+  it_behaves_like "redirects non-GDS Editors to services page", "/bad_homepage_url_status.csv"
```
(Marking resolved as positive)

KludgeKML commented on app/controllers/local_authorities_controller.rb:
```diff
@@ -3,6 +3,8 @@ class LocalAuthoritiesController < ApplicationController
 
   before_action :set_authority, except: %i[index bad_homepage_url_and_status_csv]
+  before_action :redirect_unless_gds_editor, except: %i[update download_links_csv upload_links_csv]
+  before_action :forbid_unless_gds_editor, only: %i[update download_links_csv upload_links_csv]
```
What I've done is just been explicit about everything. There's an argument that we should do:
```
before_action blah_permission, only: [ etc]
before_action blah_permission 2, only: [ etc_a etc_b]
before_action blah_permission 3, except [listed above]
```

...because it means that if someone adds an action it always gets the final permission. But we don't add actions that frequently, so it might be better to only use `only` and always just list out all actions they apply to.

(I've also fixed the download to make it forbidden)

leenagupte replied:
I think perhaps this was made more confusing by having "unless" in the action name. Things being truthy can make the logic easier to follow.
i.e. rather than `forbid_unless_gds_editor` perhaps it could be named `allow_if_gds_editor`. Then even if an `except` was used there'd only be one negative in the logic. I think it was the double negatives that were causing the brain twisting.

KludgeKML commented on app/controllers/services_controller.rb:
```diff
@@ -3,7 +3,8 @@ class ServicesController < ApplicationController
   before_action :set_service, except: :index
-  before_action :forbid_unless_permission, except: %i[index]
+  before_action :forbid_unless_permission, except: %i[index update_owner_form update_owner]
+  before_action :forbid_unless_gds_editor, only: %i[update_owner_form update_owner]
```
As above, I've made this so it just uses `only:`

leenagupte reviewed (approved):
Thanks for updating the controller permissions, I find them very easy to follow now. 🤩 
Thank you bearing with me and making the changes and for adding the extra tests.
Everything looks great👍🥇
